[#121] Update wiki-content for essential, optional and rejected features

Tuubz opened the pull request:
#121

## What happened 👀

To prevent this repository from becoming a dumping playground, we need to define some ground rules. Originally we were planning to define must-have vs nice-to-have features, but after some discussions, we realized there's a third category when the feature should not be included in the template at all. So I've revised the wording to the following:

- **Essential**
- **Optional**
- **Rejected**

In the future, we aim to include all **essential** features by default and have flags for the **optional** features, while **rejected** features are ignored. For more information, please check the code changes.

## Insight 📝

To know more about how we came to this agreement, please check the following threads:

- https://nimble-co.slack.com/archives/G41UCTGHL/p1629781993001700
- https://nimble-co.slack.com/archives/G41UCTGHL/p1629954695014500
- https://nimble-co.slack.com/archives/G41UCTGHL/p1642413515010600
- https://nimble-co.slack.com/archives/G41UCTGHL/p1642573191009000

## Proof Of Work 📹
![Screen Shot 2565-03-22 at 11 24 25](https://user-images.githubusercontent.com/18277915/159407763-58030c3c-de11-47ca-8f70-a59fdcf00d9c.png)

Tuubz commented:
Next step, code integration part for:

-  Add optional flags to the generate script command
-  Add/remove nice-to-have features to the code

I'll create an issue for it tomorrow 💯

sleepylee commented on docs/Home.md:
```diff
@@ -1,3 +1,13 @@
+
+Do _both_ apply to your feature? If not, we will define it as a **nice-to-have** feature:
+
+- Ex: [Jetpack Compose](https://developer.android.com/jetpack/compose) simplifies writing code for UI, but is not frequently used (yet)
```
it doesn't simplify, it is "a different way to write UI programmatically". The way you interpreted it makes it sounds like a nice-to-have not by itself 😅

Tuubz replied:
> Jetpack Compose is Android's modern toolkit for building native UI. It simplifies and accelerates UI development on Android bringing your apps to life with less code, powerful tools, and intuitive Kotlin APIs

I got it from the official documentation actually 😅 But I'll find another example anyways, as in the future it will be used more frequently in our projects 🚀

sleepylee commented on docs/Home.md:
```diff
@@ -1,3 +1,13 @@
+
+Planning to add a new feature? Ask yourself the following questions to define if it's a **must-have** requirement:
+
+- Is the feature frequently used in other projects? (Ex: Glide, Retrofit, ...)
```
frequently used or having the side setup that is usually verbose to replicate is also considerable.

Tuubz replied:
>  having the side setup that is usually verbose to replicate

Could you give an example of this, please? 🙇

sleepylee replied:
for example OkHttp we usually have to configure:
- Debug level
- Timeout for read/write
- Interceptor
...

Tuubz replied:
Then I would say they are frequently used. Debug level, time-outs, and interceptors are frequently used in other projects. So I think we keep it like it is, to keep it minimized. Instead, I could mention it as one of the examples for a must-have feature (with some more explanation, like we did for Lottie). What do you think? 🙏

sleepylee replied:
well...it's also good if you can specify on the setup as well; frequently used setups could be different from frequently used lib - as a lib could have many features that we don't use, so my point here is also to welcome people open PR to make configuration changes that we can make used of too...not only about adding a new lib should be considered as a must-have contribution.

Tuubz replied:
What about if we rephrase it like this?

>  Is the feature frequently used/configured in other projects?

Tuubz replied:
Or we need to avoid using "feature" as it might not clearly refer to configurations as you mentioned 🤔

sleepylee commented on docs/Home.md:
```diff
@@ -1,3 +1,13 @@
+Planning to add a new feature? Ask yourself the following questions to define if it's a **must-have** requirement:
+
+- Is the feature frequently used in other projects? (Ex: Glide, Retrofit, ...)
+- Does the feature increase productivity of the team? (Ex: Firebase, Proxyman, ...)
```
Firebase is huge, and if it's about Analysis we don't always have it.
Maybe you are mentioning Firebase Distribution or Crashlytics? Let's be explicit

sleepylee replied:
Firebase is huge, and if it's about Analysis we don't always have it.
Maybe you are mentioning Firebase Distribution or Crashlytics? Let's be explicit

sleepylee commented on docs/Home.md:
```diff
@@ -1,3 +1,13 @@
+Do _both_ apply to your feature? If not, we will define it as a **nice-to-have** feature:
+
+- Ex: [Jetpack Compose](https://developer.android.com/jetpack/compose) simplifies writing code for UI, but is not frequently used (yet)
+- Ex: [Lottie](https://lottiefiles.com/) is frequently used, but does not provide any other value unless JSON animations are needed
```
I think we can interpret it a bit more direct and less political here (it is this, but not...unless...)

How about

```suggestion
- Ex: [Lottie](https://lottiefiles.com/) a library for animation exclusively with a specific json format file. It has a specific use case and requirements that not every project is having.
```

Tuubz commented:
@Wadeewee I'll re-think about what examples to give, thanks for the suggestions 👍

Tuubz commented:
> @Tuubz It's a good idea but I'm not sure about `must-have` sometime it depend on client which is they want us to implement or depend on type of project as well because some project may be don't need API so we no need to use `Retrofit`, I feel like all of your sample are `nice-to-have`, I think for `must-have` it should be generic that not depend on feature something like these `kotlin`, `mocck`, `kotest`, `timber` What do you think? 🤔

@Wadeewee Originally, I agreed with your statement, but [after some discussions](https://nimble-co.slack.com/archives/G41UCTGHL/p1642573191009000) I've changed my mind 🙏 

It's true that the project depends on the client's requirements, but most of the time, they leave it up to us to decide which technologies to use for a new project, which we believe are best for the team and project 🚀  In terms of the Retrofit example, I still think it's an essential part. Correct me if I'm wrong, but I think it's unlikely that a project will not require an API. 🙇  

Anyhow, I've updated the content. Please have another look when available 🤩

Tuubz commented:
> @Tuubz It's a good idea but I'm not sure about must-have sometime it depend on client which is they want us to implement or depend on type of project as well because some project may be don't need API so we no need to use Retrofit, I feel like all of your sample are nice-to-have, I think for must-have it should be generic that not depend on feature something like these kotlin, mocck, kotest, timber What do you think? 🤔

@Wadeewee Originally, I agreed with your statement, but after some discussions I've changed my mind 🙏

It's true that the project depends on the client's requirements, but most of the time, they leave it up to us to decide which technologies to use for a new project, which we believe are best for the team and project 🚀 In terms of the Retrofit example, I still think it's an essential part. Correct me if I'm wrong, but I think it's unlikely that a project will not require an API. 🙇

Anyhow, I've updated the content. Please have another look when available 🤩

Wadeewee commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+- [Jacoco](https://github.com/jacoco/jacoco): Code coverage is important to maintain code quality.
+- [Hilt](https://developer.android.com/training/dependency-injection/hilt-android): Dependency injection provides a good developer experience.
+- [Firebase App Distribution](https://firebase.google.com/docs/app-distribution): Continuous delivery increases the productivity of the team.
+- [Glide](https://github.com/bumptech/glide): Image loading is always used in all the projects we've worked on.
```
@Tuubz I think guide it's an `optional` we can add it later when we have to implement feature with `image/video`  🤔 

But I think would it be good if we add `RxJava and Coroutine` in `essential` as we know we must to choose one of both before create project and what about `navigation` component, Is it essential too like `Compose` or `Hilt` What do you think 🙏

Tuubz replied:
@Wadeewee Thank you for your feedback 🚀 

> I think guide it's an optional we can add it later when we have to implement feature with image/video

In my opinion, I think `Glide` is `essential`. Because it is always used in all our previous projects. I think it's very unlikely that a future project will not have any image loading. Please correct me if I'm wrong 🙇 

> But I think would it be good if we add RxJava and Coroutine in essential as we know we must to choose one of both before create project

This is a bit of a special case, RxJava should actually be re-factored to be `optional`. While `Coroutine` should be `essential` and set as our default, as it's recommended by Google. What do you think?

> what about navigation component, Is it essential too like Compose or Hilt

Yes, the navigation component is also essential. This is because it is recommended by Google's standards 🙏

Wadeewee replied:
>This is a bit of a special case, RxJava should actually be re-factored to be optional. While Coroutine + LiveData (or Flow) should be essential and set as our default, as it's recommended by Google. What do you think?
>
>Yes, navigation component is also essential. This is because it is recommended by Google's standards 

@Tuubz I agreed with this so let add `Coroutine + Flow` and  `navigation ` in `essential` and  add  `RxJava`  in `optional` but for `LiveData` I think it old and we don't use it in `CoroutineTemplate` 🙏  and one more thing, Would it be great if we reference the example by the dependencies that we're using in our both template 🙇‍♀️

Tuubz replied:
> I agreed with this so let add Coroutine + Flow and navigation in essential and add RxJava in optional but for LiveData I think it old and we don't use it in CoroutineTemplate 🙏

For essential features, there are five criteria, so my idea is to give _one_ example for each criterion:

- I could replace the `Compose` example with `Navigation Component`, would that be better?
- Since `Coroutine` & `RxJava` are special cases and will be refactored in the future, I think it's better to avoid these examples (for now), to avoid confusion. What do you think? 🙏 

> Would it be great if we reference the example by the dependencies that we're using in our both template

Ex: Are you suggesting we should avoid using `Jacoco` as an example, because it's currently only used as a dependency for `RxJava`, but not for `Coroutine`? 🙇

Wadeewee replied:
> I could replace the Compose example with Navigation Component, would that be better?
Since Coroutine & RxJava are special cases and should be refactored in the future, I think it's better to avoid these examples (for now), to avoid confusion. What do you think? 🙏

@Tuubz Could you please explain more bit about `Since Coroutine & RxJava are special cases and should be refactored in the future` Didn't get it much 🙏 

> Ex: Are you suggesting we should avoid using Jacoco as an example, because it's currently only used in RxJava, but not for Coroutine? 🙇 I guess we could replace this example with Detekt instead 🚀
>

@Tuubz Sound good to me from now we can consider to add `Jacoco` later 🚀

Wadeewee replied:
Ah, I see @Tuubz thanks for explanation 🙇‍♀️

lydiasama commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+
+This script must include all essentials by default, while optional features can be appended with flags. To prevent this repository from becoming a dumping playground, features can be rejected too.
+
+Planning to contribute? Let's have a look at the criteria
```
```suggestion
Planning to contribute? Let's have a look at the criteria:
```

lydiasama commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+
+How do we decide if a feature is **essential**? 👍
+
+- Our team requires it to maintain code quality
+- Our team requires it to maintain a good developer experience
+- Our team requires it to increase the productivity of the team
+- It is always used or configured in the projects we've worked on
+- It is recommended to use or configure based on Google's standards
```
Shall we add a period `.` after each sentence? 🤔  Same to all bullets below.

lydiasama commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+
+**Optional**:
+- [Lottie](https://github.com/airbnb/lottie-android): Loading animations with a JSON file is regularly used, but not every project needs it.
+- [Bitrise](https://www.bitrise.io/): A CI/CD provider which is regularly used, but [GitHub Actions](https://github.com/features/actions) is our default choice.
```
I wonder about our default choice, do we include the default choice to the `essential` too?

Tuubz replied:
No, we don't need to add the default choice to `essential`, because all the essentials are our default already 🙏 

I've added the default choice here as an example, to prove that Bitrise is an `optional`

doannimble replied:
@Tuubz Imo, it would be more understandable if we separate them into: 

- `Must Have` - which cannot be missing and or replaceable.

- `Nice to have` - which can be replaceable, but so far, it is our best choice.

- `Optional` - which can be replaced by other ones anytime.

Tuubz replied:
@doannimble Please have a look at [this thread](https://nimble-co.slack.com/archives/G41UCTGHL/p1642578956011700?thread_ts=1642573191.009000&cid=G41UCTGHL) where we discuss the naming, hopefully, it can clear up the naming choice  🙏

lydiasama commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+**Essential**:
+- [Jacoco](https://github.com/jacoco/jacoco): Code coverage is important to maintain code quality.
+- [Hilt](https://developer.android.com/training/dependency-injection/hilt-android): Dependency injection provides a good developer experience.
+- [Firebase App Distribution](https://firebase.google.com/docs/app-distribution): Continuous delivery increases the productivity of the team.
```
I assume that the Firebase integration is our default choice, right? 🤔  Have you faced any projects that don't use Firebase so far? @Tuubz

Tuubz replied:
Correct, `Firebase App Distribution` is our default choice 🚀 

In my experience, the client for 7-11 chat project wanted to get the app distributed by e-mail (with help of Bitrise)

That's the only reference I have 😁

doannimble commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+- [Hilt](https://developer.android.com/training/dependency-injection/hilt-android): Dependency injection provides a good developer experience.
+- [Firebase App Distribution](https://firebase.google.com/docs/app-distribution): Continuous delivery increases the productivity of the team.
+- [Timber](https://github.com/JakeWharton/timber): A logging library which is always used in all the projects we've worked on.
+- [Navigation Component](https://developer.android.com/guide/navigation/navigation-getting-started): A suite of libraries for in-app navigation which is recommended by Google.
```
@Tuubz I think `Navigation Component` should be `Optional`, as not all projects need it 🤔

sleepylee replied:
I think of the opposite...if it is optional then you will have to provide the primary way to navigate in an application...I don't think the traditional fragment backstack manipulation is preferred right?

Tuubz replied:
@doannimble Could you perhaps give an example of which project didn't need it?

In my opinion, anything that Google recommends using should be `Essential`. So as they recommend using "[Single Activity with multiple fragments](https://youtu.be/9O1D_Ytk0xg?t=79)", navigation component comes with.

doannimble replied:
@sleepylee @Tuubz  We can use fragment backstack with activity if it works and does not give us an issue. You guys can see at the end of the video [5:30](https://www.youtube.com/watch?v=9O1D_Ytk0xg&t=330s)  that if you're using Multi-Activity development and it works, stick with it. 

But if you're running into issues and bugs then to use Single Activity development.

The old way is not to be preferred, but it does not mean the `Navigation Component` is a must-have IMO 🙏

sleepylee replied:
we cannot leave the place for doubt like that... please read the description again on the essential part: the feature targetted as default or essential is the one that increase our dev experience and productivity. This is not a playground or fully crowd open source that we have to leave space for interpretation. We must own it and make the solid rule in the end.
As you can see, every project is now moving to navigation component; will you start the next project without it? Just answer that question to yourself then you might have the clear answer.

lydiasama replied:
@doannimble As for the android-template is for `us` to use it to initiate the android project that we're gonna work on it further. So I think this should combine the approach we usually use in the client project. 🙏🏻  So I think the `Navigation Component` is an Essential one. What do you think?

doannimble replied:
Got it, thank you @Tuubz @sleepylee @lydiasama 🙏

sleepylee commented:
@doannimble this proposal has a few flaws that you are leaving for further confusion:

> Must Have - which cannot be missing and or replaceable.

there is nothing `not replaceable` forever...so it's either the only option for now (navigation component), or there are 2 ways to do things but 1 must be the default (coroutine over rxjava).

> Nice to have - which can be replaceable, but so far, it is our best choice.

Now when you read again, you actually reference ` but so far, it is our best choice.` to the first choice

> Optional - which can be replaced by other ones anytime.

and lastly, this same repetition `which can be replaced` will create confusion to categorize type with type 2. 


Please understand the idea first, having an opinion and cross-check on your understanding before proposing a new idea.

sleepylee commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+
+How do we decide if a feature is **essential**? 👍
+
+- Our team requires it to maintain code quality.
+- Our team requires it to maintain a good developer experience.
+- Our team requires it to increase the productivity of the team.
+- It is always used or configured in the projects we've worked on.
+- It is recommended to use or configure based on Google's standards.
```
I think we can reorder this a bit so it has a parallelism structure with the other option below. What I meant is that the first consideration must be about `how often it has been chosen or agreed to be picked` in a project.

The first 3 can be shortened to reduce the repetition in the writing, be precise.
>- Our team requires it to maintain code quality.
>- Our team requires it to maintain a good developer experience.
>- Our team requires it to increase the productivity of the team.

lastly, suggesting a more directive way of writing:
> It is a Google recommended component.

Tuubz replied:
@sleepylee Yes, there is some repetition: `Our team requires it to...` 

Do you suggest minimizing it in one sentence? I'm open to hearing your approach 🙏

sleepylee replied:
yes, I'd say we write it in a crystal-clear thought:
something like:


> - It is always implemented in the projects we have worked on or will work on.
> - It allows developers to increase the productivity from boilerplate setup, maintain the quality, and have a good experience for themselves.
>- It is a Google-recommended component.

what do you think?

Tuubz replied:
Thanks for the recommendations, I agree with the first and third bullet points.

For the second bullet point, we can't only refer to `developers`. 

Ex: `Chucker` or `Firebase App Distribution` also increases the productivity of our project managers.

sleepylee replied:
developers have the most benefit from it, in this codebase PM comes after so I would refrain to prioritize them in this topic, it will generate more conflict to resolve later when someone wants to bring something that has less value to dev but more value to the PM (btw, Product Manager, we don't have Project Manager role), how do you think?

Tuubz replied:
If I'm not mistaken, it was Julien that requested for Chucker.

As for Firebase App Distribution, is mostly beneficial for testers (our PM).

Please correct me if I got the wrong idea here 🙏

sleepylee replied:
nope, it was from our side to initialize Chucker after we learned from iOS. Eventually, we found it very useful for dev ourselves as well.

For Firebase App Distribution, it is to enhance our productivity because it's a requirement that often appears in our process, so we want to reduce the setup time. Again, the purpose of the feature is for everyone, the template is here to serve us: save our time on the setup.

Tuubz replied:
I agree, that it saves _our_ time on the setup. But both features were implemented for the team's purpose:

- We can check LogCat, so Chucker is not essential for us developers. It is essential for PMs however
- PMs require Firebase App Distribution for testing. So it's for their needs that we set it up 

🙏

sleepylee replied:
I still don't get your point from the latest argument. As been said, I don't deny that PM benefits from it, but I suggested to not involving them in this context because it can generate more debates in the future.

Now talking about Firebase or Chucker: 
- Do developers use them? Yes.
- Do we want to save time to set them up? Yes

So can we just talk about developers in here? I think yes too?

sleepylee replied:
> We can check LogCat, so Chucker is not essential for us developers. It is essential for PMs however.

LogCat is a console display, while Chucker allows you to debug while not connecting to AS. I guess you haven't had this experience before in AllOnline but we did have some good use cases with it in DM where we started using it more often. While it is not a blocker for our development, it has proved its usefulness to us, as well as the PM, so that's why it got promoted here to reduce the setup time.

sleepylee replied:
For firebase: to shape the understanding of "developer-oriented" way: it is a requirement that often come from the PM; while it benefits the PMs, we don't necessarily bring PM here as the reason. The real reason or the real benefit **to the developers** is that we don't have to re-setup it, again and again, all the time. Does this sound different from your original thought? 🤔

sleepylee commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+How do we decide if a feature is **optional**? 🚩
+
+- It is regularly used or configured in projects we've worked on.
+- It is regularly used or configured as an alternative to the default choice.
```
recommend avoiding repetition. This can be understood simply as: 
- We used it too, but maybe now there is a newer dominant option (coroutine over rxjava so rxjava becomes optional).

sleepylee commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+How do we decide if a feature is **rejected**? 👎
+
+- It is barely used or configured in projects we've worked on.
+- It is barely used or configured as an alternative to the default choice.
```
same here, we need to be more directive and succinct: Components that we don't apply to most of the project, come as a specific project requirement. Right? 🤔

sleepylee replied:
and the idea here is to unify the ways/components our team should be familiar with and spend effort on maintaining, so by rejecting the uncommon idea, we can have more focus on the real matter.

minhnimble commented on docs/Home.md:
```diff
@@ -1,3 +1,50 @@
+
+Keep in mind, the features are based on _our team's_ requirements. In case the client has different requirements or requests, we can consider adding them as optional features if they occur regularly.
+
+Know that the above examples are not definitive as new and existing libraries keep on emerging and evolving.
```
Using the word `Know that` here doesn't seem to be appropriate for the given context, I suggest to update it to `Please note that` for better wording on politely letting the readers be aware of what is going on. What do you think?
```suggestion
Please note that the above examples are not definitive as new and existing libraries keep on emerging and evolving.
```

Tuubz commented:
@minhnimble @lydiasama @Wadeewee @sleepylee Updated 🚀

minhnimble reviewed (approved):
LGTM 🚀

luongvo commented on docs/Home.md:
```diff
@@ -1,3 +1,48 @@
+- [Navigation Component](https://developer.android.com/guide/navigation/navigation-getting-started): A suite of libraries for in-app navigation which is recommended by Google.
+
+**Optional**:
+- [Lottie](https://github.com/airbnb/lottie-android): Loading animations with a JSON file is regularly used, but not every project needs it.
```
This could be considered as `Rejected`, the same as `Skeleton Layout` too IMO. Instead of mentioning `Lottie` as `Optional`, I think we can list `Glide - an image loader` here because almost project needs an image loader, but not always `Glide` 🤓

sleepylee replied:
>I think we can list Glide - an image loader here because almost project needs an image loader, but not always Glide 

@luongvo do you mean

>....but not always `Lottie`
?

because 10/10 recent projects we had we always use Glide 🤔

luongvo replied:
@sleepylee Okay if 10/10 recent projects we had we always use Glide so Glide is an essential feature, I may be wrong on this part.

On another side, I still don't think we can consider `Lottie` as Optional, it should be the same as `Skeleton Layout` in example.

Tuubz replied:
@luongvo At this moment, I think I have used `Lottie` in at least of half of my projects, while for `Skeleton` it's zero.

Anyways, I'm totally open to finding a more clear example. Do you have any suggestion? 🙇

sleepylee replied:
@luongvo yep probably we can reject it too as Lottie's setup is relatively simple with a library import. Possibly for those cases, depending on the project's need - developers should setup themselves instead of us maintaining it.

sleepylee replied:
@Tuubz I think we have Skeleton loading view in The 1 + DeeMoney (not sure, maybe Toyota Wallet also) 🤔

Tuubz replied:
I'll revise the rules between "optional" & "rejected" a bit more, so it's more clear and strict + adapt the examples accordingly 👍

luongvo replied:
@Tuubz @sleepylee How about listing [RxPermissions](https://github.com/tbruyelle/RxPermissions) as an Optional feature here instead? I see most of our projects use it and it's quite convenience for our permission handling steps too 🤔

sleepylee replied:
@luongvo that works 👍

luongvo reviewed (commented):
Rest lgtm

Tuubz commented:
@luongvo @sleepylee @lydiasama @Wadeewee Updated 🤩

luongvo reviewed (approved):
All good, LGTM

Tuubz commented:
@doannimble @sleepylee Any final words? Please do, if any 😁

doannimble reviewed (approved):
LGTM 🚀
